.expeditor/scripts/release_habitat/package_and_upload_binary.sh

```diff
@@ -0,0 +1,126 @@
+if [[ $BUILD_PKG_TARGET == *"darwin" ]]; then
+  artifact_name=$(buildkite-agent meta-data get MACOS_ARTIFACT)
+  buildkite-agent artifact download "$artifact_name" "$tmp_root/" --step "[:macos: build hab]"
+  mv "$tmp_root/$artifact_name" "$tmp_root/hab-$channel.hart"
+else
+  hab pkg exec core/wget wget "${HAB_BLDR_URL}/v1/depot/pkgs/core/hab/$release_version/download?target=$BUILD_PKG_TARGET" -O "$tmp_root/hab-$channel.hart"
+fi
+
+target_hart="$tmp_root/hab-$channel.hart"
+tail -n+6 "${target_hart}" | \
+    tar --directory "${extract_dir}" \
+        --extract \
+        --xz \
+        --strip-components=6
+
+extracted_hab_binary="$(find "$extract_dir" \( -name hab -or -name hab.exe \) -type f)"
+pkg_target="$(tr --delete '\r' < "${extract_dir}"/TARGET)"
+pkg_arch="$(echo "$pkg_target" | cut -d '-' -f 1)"
+pkg_kernel="$(echo "$pkg_target" | cut -d '-' -f 2,3)"
+pkg_ident="$(tr --delete '\r' < "$extract_dir"/IDENT)"
+pkg_origin="$(echo "$pkg_ident" | cut -d '/' -f 1)"
+pkg_name="$(echo "$pkg_ident" | cut -d '/' -f 2)"
+pkg_version="$(echo "$pkg_ident" | cut -d '/' -f 3)"
+pkg_release="$(echo "$pkg_ident" | cut -d '/' -f 4)"
+archive_name="hab-$(echo "$pkg_ident" | cut -d '/' -f 3-4 | tr '/' '-')-$pkg_target"
+
+build_dir="$tmp_root/build"
+pkg_dir="$build_dir/${archive_name}"
+
+echo "Copying $extracted_hab_binary to $(basename "$pkg_dir")"
+mkdir -p "$pkg_dir"
+mkdir -p "$tmp_root/results"
+
+if [[ $pkg_target == *"windows" ]]; then
+for file in "$(dirname "$extracted_hab_binary")"/*; do 
+  cp -p "$file" "$pkg_dir/"
+done
+else
+  cp -p "$extracted_hab_binary" "$pkg_dir/$(basename "$extracted_hab_binary")"
+fi
+
+echo "Compressing 'hab' binary"
+pushd "$build_dir" >/dev/null
+case "$pkg_target" in
+*-linux | *-linux-kernel2)
+    pkg_artifact="$tmp_root/results/${archive_name}.tar.gz"
+    tarball="$build_dir/$(basename "${pkg_artifact%.gz}")"
+    hab pkg exec core/tar tar cf "$tarball" "$(basename "$pkg_dir")"
+    rm -fv "$pkg_artifact"
+    hab pkg exec core/gzip gzip -9 -c "$tarball" > "$pkg_artifact"
+    ;;
+*-darwin | *-windows)
+    pkg_artifact="$tmp_root/results/${archive_name}.zip"
+    rm -fv "$pkg_artifact"
+    hab pkg exec core/zip zip -9 -r "$pkg_artifact" "$(basename "$pkg_dir")"
+    ;;
+*)
+    exit_with "$target_hart has unknown TARGET=$pkg_target" 3
+    ;;
+esac
+
+# Generate our shasum
+popd >/dev/null
+pushd "$(dirname "$pkg_artifact")" >/dev/null
+sha256sum "$(basename "$pkg_artifact")" > "${pkg_artifact}.sha256sum"
+
+# Sign our artifact
+popd >/dev/null
+pushd "$(dirname "$pkg_artifact")" >/dev/null
+gpg --armor --digest-algo sha256 --default-key 2940ABA983EF826A --output "$(basename "$pkg_artifact").asc" --detach-sign "$(basename "$pkg_artifact")"
+popd
+
+# Name of the file to upload
+upload_artifact="$(basename "$pkg_artifact")"
+# Now strip the `release` so we have a 'latest'
+latest_artifact="$(basename "$pkg_artifact" | sed -E 's/[0-9]{14}/latest/')"
```

Comment:
I'd rather have that fail at this point. To cm's point - that would required a pretty big reshuffle of the order of code that could introduce more complications at this point.

Reply:
Would it fail cleanly, though? If we were do add more digits to the `release` component (something we've talked about), this would result in `latest_artifact` having something line `latest123` it it.

Reply:
I guess my question is - how likely would it be to change the datestamp for the release? that seems like a pretty fundamental thing that we would not change.

Reply:
We've been actively discussing it as a way to remediate the several times we've had collisions. I'm not aware of the current format being formally specified anywhere, so I think it definitely *could* change.

I don't think this code need to be maximally general or anything, but validating that what we find *is* the format we expect and producing a clear error otherwise would be more robust than assuming that format is inviolable.

Reply:
I think it would be good to address this in some way